update-v1.c

```diff
@@ -0,0 +1,376 @@
+#define SUPER_FL_BLOCK_CRC 65097 // 0xFE49
+#define SUPER_FL_FLASH_CMD 65098 // 0xFE4A
+#define SUPER_FL_FLASH_STS 65099 // 0xFE4B
+#define SUPER_FL_BLOCK_DATA_LEN 64
+
+/* Read-back status values */
+/* Default value of status, closed */
+#define STATUS_CLOSED 0x00
+/* Once the flashwrite process is set up, but no data written */
+#define STATUS_READY 0xAA
+/* Flashwrite process has seen full length of data written and is considered done */
+#define STATUS_DONE 0x01
+/* Flashwrite is in process, meaning SOME data has been written, but not the full length */
+#define STATUS_IN_PROC 0x02
+/* A CRC error occurred at ANY point during data write. Note that this status
+ * is not set if CRC fails for open process, the system simply does not open
+ */
+#define STATUS_CRC_ERR 0x03
+/* An error occurred while trying to erase the actual flash */
+#define STATUS_ERASE_ERR 0x04
+/* An error occurred at ANY point during data write. */
+#define STATUS_WRITE_ERR 0x05
+/* Erase was successful, but, the area to be written was not blank */
+#define STATUS_NOT_BLANK 0x06
+/* A BSP error opening and closing flash. Most errors are buggy code, configurations, or unrecoverable */
+#define STATUS_OPEN_ERR 0x07
+/* Wait state while processing a write */
+#define STATUS_WAIT 0x08
+/* Request the uC reboot at any time after its open status */
+#define STATUS_RESET 0x55
+
+enum super_flash_status {
+	SUPER_UPDATE_ON_REBOOT = (1 << 8), /* Set when the APPLY_REBOOT command is issued */
+	/* Bits 7:0 are STATUS_ from flashwrite */
+};
+
+enum super_flash_cmd {
+	SUPER_APPLY_REBOOT = (1 << 3),
+	SUPER_CLOSE_FLASH = (1 << 2),
+	SUPER_OPEN_FLASH = (1 << 1),
+	SUPER_WRITE_BLOCK = (1 << 0),
+};
+
+/* Some return values of tend() */
+enum i2c_cmds_t {
+	I2C_NOCMD = (0 << 0),
+	I2C_REBOOT = (1 << 0),
+	I2C_HALT = (1 << 1),
+};
+
+enum gen_flags_t {
+	GEN_FLAG_LED_DAT = (1 << 3),
+	GEN_FLAG_OVERRIDE_LED = (1 << 2),
+	GEN_FLAG_WAKE_EN = (1 << 1),
+	GEN_FLAG_ALARM_TYPE = (1 << 0),
+};
+
+enum gen_inputs_t {
+	GEN_INPUTS_USB_VBUS = (1 << 1),
+	GEN_INPUTS_EN_DB9_CONSOLE = (1 << 0),
+};
+
+enum super_features_t {
+	SUPER_FEAT_SN = (1 << 2),
+	SUPER_FEAT_FWUPD = (1 << 1),
+	SUPER_FEAT_RSTC = (1 << 0),
+};
+
+struct micro_update_footer_v1 {
+	uint32_t bin_size;
+	uint16_t revision;
+	uint8_t flags;
+	uint8_t misc;
+	uint16_t model;
+	uint8_t footer_version;
+	uint8_t magic[11];
+} __attribute__((packed));
+
+void flash_print_error(uint8_t status)
+{
+	switch (status) {
+	case STATUS_OPEN_ERR:
+		fprintf(stderr, "Flash failed to open!\n");
+		break;
+	case STATUS_NOT_BLANK:
+		fprintf(stderr, "Flash not blank\n");
+		break;
+	case STATUS_ERASE_ERR:
+		fprintf(stderr, "Flash failed to erase!\n");
+		break;
+	case STATUS_WRITE_ERR:
+		fprintf(stderr, "Flash failed to write!\n");
+		break;
+	case STATUS_CRC_ERR:
+		fprintf(stderr, "Flash received bad data CRC!\n");
+		break;
+	default:
+		fprintf(stderr, "Unknown flash failure\n");
+		break;
+	}
+}
```

Comment:
It looks like this feedback/error printing isn't present in v0 update

Reply:
Fixed, added a update-shared.{c,h}